### PR TITLE
[inductor] Fix torch.split bug on unbacked symint

### PR DESCRIPTION
**Pull request by andrewlee302**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113406

torch.split(x, l) fails when l's shape is the unbacked symint.

E.g. l =
y.tolist() makes l the unbacked shape, because l depends on the
data access of y. The downdtream call `SliceView.create()`
evaluates the shape even if the input shape is unbacked symint,
which brings up the bug.

Test Plan:
python test/inductor/test_unbacked_symints.py -k test_split_with_sizes

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by aakhundov — APPROVED** [review]
LGTM, thanks @andrewlee302! Please fix EasyCLA and we're good to go.

File: torch/_inductor/ir.py
Comment (peterbell10) [inline torch/_inductor/ir.py:2088]: Isn't this necessary for correctness? I think the correct fix would be to use `sympy.Min` instead of `evaluate_min`.

File: torch/_inductor/ir.py
Comment (peterbell10) [inline torch/_inductor/ir.py:2091]: This part looks like an optimization that we can safely drop.

**Review by peterbell10 — CHANGES_REQUESTED** [review]

File: torch/_inductor/ir.py
Comment (aakhundov) [reply torch/_inductor/ir.py:2088]: Thanks for the heads-up! Yeah, these lines seem relevant for unbacked SymInts, too (although not sure how well will `sympy.Min`'s application work downstream). @andrewlee302 please take a look, thanks!

Comment (andrewlee302) [reply torch/_inductor/ir.py:2088]: `Sympy.Min` replacement doesn't work well here. Briefly, codegen fails. Detailed logs below:
```
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL] Error in codegen for ComputedBuffer(name='buf3', layout=FixedLayout('cuda', torch.float32, size=[], stride=[]), data=Reduction(
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   'cuda',
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   torch.float32,
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   def inner_fn(index, rindex):
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]       r0 = rindex
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]       tmp0 = ops.load(arg1_1, r0 + Min(0, i9))
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]       return tmp0
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   ,
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   ranges=[],
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   reduction_ranges=[-Min(0, i9) + Min(32, i9)],
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   reduction_type=sum,
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   origin_node=sum_1,
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   origins={sum_1}
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL] ))
```

I think here are two possible implications that we need to figure it out:
1. This short circuiting works well for backed symint. Is it necessary to need it and what's the benefits for the quick return? As @ezyang introduced it [here](https://github.com/pytorch/torchdynamo/commit/2788fccfc6142fe2bf1b128d9f1001402d90886e#diff-194018e9855f41edb48ecad770b7b3aa6439f6237f640d273bf1e3aa0e8377ba) and he may answer it.
2. Sympy on unbacked symint makes the codegen not work. Is it expected from the workflow perspective or sympy should be well supported for resolving this fix? 

I think both questions can be answered and this fix should be reasonable for now. We can move further later. @adnanaziz  @ezyang @peterbell10 WDTY?

File: torch/_inductor/ir.py
Comment (andrewlee302) [reply torch/_inductor/ir.py:2091]: Covered in above comment.

File: torch/_inductor/ir.py
Comment (aakhundov) [reply torch/_inductor/ir.py:2088]: @andrewlee302 I think, the short-circuiting part is not very important: we can either keep it behind newly-added `if` statement checking that there are no unbacked SymInts in the `start` and `end`. Or, as @peterbell10 suggested, we can also drop it.

Regarding your error, I think, we may want to include unbacked SymInts (i.e., "i") in the tuple [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py#L1145-L1146) to circumvent them same as we currently circumvent backed SymInts (i.e., "s"). @ezyang does this sound sound to you?

But then we hit another error during compilation of the generated Triton code:

```
...
  File "/data/users/aakhundov/pytorch/torch/_inductor/codecache.py", line 887, in _run_from_cache
    return compiled_graph.compiled_artifact(inputs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/torchinductor_aakhundov/r4/cr4vabmhar475qxxdzxs5bd5tgi5plnogjqe2ngloi4b36p4cxy5.py", line 89, in call
    triton_red_fused_sum_0_rnumel = ((-1)*(Min(0, i9))) + (Min(32, i9))
                                           ^^^
NameError: name 'Min' is not defined
```

Seems that `sympy.Min` is not properly rendered in the Python wrapper codegen. @andrewlee302 I'd try adding min to the `PythonPrinter` [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/common.py#L340).

**Review by aakhundov — CHANGES_REQUESTED** [review]
Please address the items from the last comment. Thanks!

File: torch/_inductor/ir.py
Comment (ezyang) [reply torch/_inductor/ir.py:2088]: So, the first thing to figure out, is what we actually want the code to be generated in this case to actually have a min/max. I find that often I reach for truly symbolic min/max when actually I can infer that a guard should go one way or another at a higher level. This is the preferred resolution and sidesteps the codegen problems here.

Looking at the tests, it seems to me that it should be statically obvious that you are within start/end. So it seems to me that we should be able to do some `evaluate_static_expr` business to get True/False without stubbing our toe on guard on unbacked symint. For examples of me having done similar refinements in the past, check https://github.com/pytorch/pytorch/pull/110979 https://github.com/pytorch/pytorch/pull/108685

**Comment by ezyang** [comment]
bump

File: torch/_inductor/ir.py
Comment (peterbell10) [reply torch/_inductor/ir.py:2088]:
```python
    triton_red_fused_sum_0_rnumel = ((-1)*(Min(0, i9))) + (Min(32, i9))
```

`Min(0, i9)` probably shouldn't be there, sympy should be able to simplify it away to 0 if the symbol `i9` was created with the `nonnegative=True` assumption.

`Min(32, i9)` on the other hand is more tricky, we should be able to do some range analysis to bound `i9` but in the completely general case I think we *do* want to codegen a `min` here.

Comment (ezyang) [reply torch/_inductor/ir.py:2088]: I don't buy it. If the slice is in bounds, the result here should always be i9. Because the slices are coming from a torch.split, we know they are guaranteed to be in bounds after we've done error checking. The error checking itself must be done at runtime, but by the time we get to the kernel we know they must have passed.

(BTW: we aren't actually emitting the deferred runtime asserts, this is still on my list of things to do.)

Comment (aakhundov) [reply torch/_inductor/ir.py:2088]: @ezyang to clarify, the codegened `Min` is not incorrect in this case (modulo the required extension of the `PythonPrinter`), just redundant? As I understand, `torch.split` does range checks and, due to those, in the generated code we can be sure that the split indices are in-range. But the deferred runtime asserts are not emitted currently. So would it make sense to keep the redundant `Min`s codegened for now and land this PR to unblock; and then remove redundant `Min`s in a follow-up, when we start emitting runtime asserts?

Comment (ezyang) [reply torch/_inductor/ir.py:2088]: The codegenned min is "correct", but it is unnecessary.

You are welcome to unblock this by putting in min codegen, if you can get it to work. But if it seems difficult to do (which it does seem like it is causing some trouble), maybe better to deal with the root cause rather than bashing through it.

Comment (andrewlee302) [reply torch/_inductor/ir.py:2088]: Thanks for the great input. I'm very new to pytorch and trying to interpret the discussion. Please correct me if I was wrong.
1. Dynamo does the error check for `split_with_sizes` op ([code pointer](https://github.com/pytorch/pytorch/blob/9b0f2f8d94b7485c2c2a4e3d1294a3589b3466b0/torch/_decomp/decompositions.py#L1221-L1248) before triton kernel codegen. I feed `torch.split(x, [35,-3])` and `torch.split(x, [32])` to the test and it proves that the error check happens in step 1. (logged as `[INFO] Step 1: torchdynamo start tracing`). 
2. Due to #1 that assert the range should be valid, the downstream calculation is redundant. E.g. triton code `triton_red_fused_sum_0_rnumel = ((-1)*(Min(0, i9))) + (Min(32, i9))` can be theoretically equivalent to `triton_red_fused_sum_0_rnumel = 32`.
3. But triton python code is problematic due to the lack of `from sympy import Min`. To make it runnable, we need to refine the triton codegen component. My quick question here: how difficult it is?
4. The [sizevars.evaluate_min](https://github.com/pytorch/pytorch/blob/9b0f2f8d94b7485c2c2a4e3d1294a3589b3466b0/torch/_inductor/ir.py#L2087-L2088) within `SliceView.create` happens after error handling for `torch.split` op. If all ops depending on `SliceView` can safeguard the assertion, looks like we can remove `sizevars.evaluate_min` sentences. Can we?
5. The short-circuiting is not a necessary for functionality or performance. We should get rid of it to simplify the code.

@ezyang @adnanaziz @peterbell10 WDYT?

**Comment by andrewlee302** [comment]
PR Update: remove the short-circuiting and start/end alignment in `SliceView.create`, to see if CI tests could be broken.

File: torch/_inductor/ir.py
Comment (aakhundov) [reply torch/_inductor/ir.py:2088]: @ezyang Thanks for being flexible :) I think, the only thing that is missing to make the codegen-ed `sympy.Min` work is a small extension of the [`PythonPrinter`](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/common.py#L340) to add `_print_Min` (similar to the existing [`_print_Max`](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/common.py#L371-L373)). This makes @andrewlee302's unit test pass for me even when `sizevars.evaluate_min` calls are outside the new `if` statement.

@andrewlee302 IIUC, 1. and 2. sound about right. For 3., it's easy: please see the above paragraph / give it a try in the PR. For 4., I don't think we can remove `sizevars.evaluate_min` completely, as (IIUC) the runtime assertions are not yet generated from the `torch._check*`s; you can also see lots of unit tests failing now as you've removed the code in question. For 5., this is what @peterbell10 has suggested, yeah.

Comment (ezyang) [reply torch/_inductor/ir.py:2088]: yes, just update python printer

**Review by ezyang — APPROVED** [review]
slick

**Comment by jansel** [comment]
@aakhundov can you take another look?

**Comment by aakhundov** [comment]
> @aakhundov can you take another look?

@jansel thanks for the heads up. We're in offline contact with @andrewlee302. There are some e2e model accuracy issues that need to be investigated and resolved before landing this PR.

File: torch/fx/experimental/validator.py
Comment (aakhundov) [inline torch/fx/experimental/validator.py:366]: @andrewlee302 Could you briefly explain why do we need to change this? Thanks!

Comment (andrewlee302) [reply torch/fx/experimental/validator.py:366]: Update the commit, to address torchbench error on some models. Log failure snippet ([raw log](https://ossci-raw-job-status.s3.amazonaws.com/log/18915848660)):
```
2023-11-22T04:23:29.6156040Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/fx/experimental/validator.py", line 372, in __getattr__
2023-11-22T04:23:29.6156964Z     raise AttributeError(f"unhandled operator: {name}")
2023-11-22T04:23:29.6157656Z torch._dynamo.exc.InternalTorchDynamoError: unhandled operator: minimum
```

Local test passes. A command to repro is:
```
python benchmarks/dynamo/torchbench.py --ci --accuracy --timing --explain --inductor --device cpu --inference --float32 --total-partitions 1 --partition-id 0 --only vision_maskrcnn
```

Comment (aakhundov) [reply torch/fx/experimental/validator.py:366]: I'd guess it somehow being related to adding `sympy.Min` to the IR in this PR. @ezyang does this change look good to you? I believe, this just adds a key to the pre-existing `REPLACEMENT` dict which isn't likely to disrupt anything?

Comment (ezyang) [reply torch/fx/experimental/validator.py:366]: yeah, this should be fine. Maybe do maximum too?

**Comment by aakhundov** [comment]
Seems that the two failing CI jobs are failing consistently with this constraint violation error (and a bunch of CUDA OOMs):

```
  - Not all values of RelaxedUnspecConstraint(L['inputs'][0].size()[0]) are valid because L['inputs'][0].size()[0] was inferred to be a constant (4).
```

I'm wondering if / how this error can be triggered by the changes in this PR. @ezyang any thoughts? Thanks!

File: torch/_inductor/sizevars.py
Comment (aakhundov) [inline torch/_inductor/sizevars.py:347]: @andrewlee302 I support the idea, but the implementation looks a bit intrusive: `evaluate_min` may be used in many other places and it's hard to be sure that the suggested workaround would work there, too (I think, it would be cleaner to let those places discover themselves explicitly with a more clear error). So I'd suggest moving this condition one level above to the `SliceView.create` to address the case in question specifically without changing the `evaluate_min` code. The effect on this PR should be the same.

**Review by aakhundov — CHANGES_REQUESTED** [review]

**Review by aakhundov — APPROVED** [review]
Thanks @andrewlee302!

**Comment by DanilBaibak** [comment]
@andrewlee302 sorry for reverting your change. There are a bunch of inductor failures and we suspect that this may happen due to your changes. Trying to revert to see if it helps fix the issue.

**Comment by andrewlee302** [comment]
@DanilBaibak Could you please share the failures for reproducing?

**Comment by DanilBaibak** [comment]
@andrewlee302 the problem happen not due to your changes. You may need to rebase your PR and if the tests pass, you can merge it.

File: torch/_inductor/codegen/triton.py
Comment (jansel) [inline torch/_inductor/codegen/triton.py:1147]: Can you explain why this is needed?

Comment (aakhundov) [reply torch/_inductor/codegen/triton.py:1147]: From what I remember, the earlier version of this PR has run into an error. There are more details in [this thread](https://github.com/pytorch/pytorch/pull/113406#discussion_r1390927602). cc @andrewlee302 for more details. Thanks!

**Pull request closed without merging**
